src/api-service/__app__/onefuzzlib/agent_events.py

```diff
@@ -0,0 +1,239 @@
+    if not node:
+        return Error(code=ErrorCode.INVALID_NODE, errors=["unable to find node"])
+    return node
+
+
+def on_state_update(
+    machine_id: UUID,
+    state_update: NodeStateUpdate,
+) -> Optional[Error]:
+    state = state_update.state
+    node = get_node(machine_id)
+    if isinstance(node, Error):
+        return node
+
+    if state == NodeState.free:
+        if node.reimage_requested or node.delete_requested:
+            logging.info("stopping free node with reset flags: %s", node.machine_id)
+            node.stop()
+            return None
+
+        if node.could_shrink_scaleset():
+            logging.info("stopping free node to resize scaleset: %s", node.machine_id)
+            node.set_halt()
+            return None
+
+    if state == NodeState.init:
+        if node.delete_requested:
+            logging.info("stopping node (init and delete_requested): %s", machine_id)
+            node.stop()
+            return None
+        node.reimage_requested = False
+        node.state = state
+        node.save()
+        return None
+
+    logging.info("node state update: %s from:%s to:%s", machine_id, node.state, state)
+    node.state = state
+    node.save()
+
+    if state == NodeState.setting_up:
+        # Model-validated.
+        #
+        # This field will be required in the future.
+        # For now, it is optional for back compat.
+        setting_up_data = cast(
+            Optional[NodeSettingUpEventData],
+            state_update.data,
+        )
+
+        if setting_up_data:
+            if not setting_up_data.tasks:
+                raise Exception("setup without tasks.  machine_id: %s", machine_id)
+
+            for task_id in setting_up_data.tasks:
+                task = Task.get_by_task_id(task_id)
+                if isinstance(task, Error):
+                    return task
+
+                # The task state may be `running` if it has `vm_count` > 1, and
+                # another node is concurrently executing the task. If so, leave
+                # the state as-is, to represent the max progress made.
+                #
+                # Other states we would want to preserve are excluded by the
+                # outermost conditional check.
+                if task.state != TaskState.running:
+                    task.state = TaskState.setting_up
+
+                task.on_start()
+                task.save()
+
+                # Note: we set the node task state to `setting_up`, even though
+                # the task itself may be `running`.
+                node_task = NodeTasks(
+                    machine_id=machine_id,
+                    task_id=task_id,
+                    state=NodeTaskState.setting_up,
+                )
+                node_task.save()
+
+    elif state == NodeState.done:
+        # if tasks are running on the node when it reports as Done
+        # those are stopped early
+        node.mark_tasks_stopped_early()
```

Comment:
Double-checking: this only impacts _running_ tasks?

Reply:
Also, this seems exceptional and a sign of a bug. The node is done, but there are tasks running. Does this imply error logging? Is there a normal path where this could happen that I'm not thinking of?

Reply:
Yes, this only impacts running tasks, though the logic is a bit convoluted.  Refactoring that to be more clear is out of scope for what this PR is focused on.

mark_tasks_stopped_early iterates over the NodeTasks, calling `task.mark_failed()` for each task.  `Task.mark_failed` checks if the task state is already in Stopping/Stopped.